src/core/net/srp_server.cpp

```diff
@@ -323,7 +323,7 @@ bool Server::HasNameConflictsWith(Host &aHost) const
         {
             if (host.HasServiceInstance(service.GetInstanceName()))
             {
-                VerifyOrExit(*aHost.GetKey() == *host.GetKey(), hasConflicts = true);
+                VerifyOrExit(aHost.GetKey()->GetKey() == host.GetKey()->GetKey(), hasConflicts = true);
```

Comment:
One idea that may work and may make it simpler: Instead of remembering `Dns::Ecdsa256KeyRecord` in `Host`, we can consider remembering the `PublicKey`  along with a boolean `mIsKeyValid`.

Reply:
@abtink 
Instead of using a `mIsKeyValid` boolean, what about making the key a `RetainedPtr`?

Reply:
Good idea (basically using a pointer to indicate presence of the key). I think an `OwnedPtr` may work better since there should be only instance of each key maintained by a `Host` (which should be freed when `Host` instance is freed).

Reply:
@abtink I agree there is no necessary to save the entire KEY RR and it will better to use the smart pointers or `Heap::Data` constructions. While it seems the retain_ptr and owned_ptr require a `Free()` function in the type and `Heap::data` will require `static_cast` between `Crypto::Ecdsa::P256::PublicKey` and `uint_8 *`. So I am going to keep this change as is and small. I also left a TODO comment for thinking about it later. Are you okay with that?

Reply:
`RetainPtr` may not be a good fit for this case? Per my understanding, `RetainPtr` provides a way to hold a refernce to a resource, which uses `Free()` and `IncrementRetainCount()` for reference counting. So, this means the resource should be potentially owned by multiple owners, which however is not the case here.

Reply:
> So I am going to keep this change as is and small. I also left a TODO comment for thinking about it later. Are you okay with that?

SGTM. I think would be good to add this fix in. We can consider other enhancments later.

If we use `Heap::Data mKey`  to save the key, I think we can populate it using something like `mKey.SetFrom(publicKey.GetBytes(), Crypto::Ecdsa::PublicKey::kSize)`. 


> What we need here is something like std::unique_ptr, but we don't have such support in OpenThread core yet.

We do. `OwnedPtr<>` is similar to `std::unique_ptr`. 😃

Reply:
Just adding a quick note on `RetainPtr`: If we have a type `T` and want to use it in `RetainPtr` we can create a sub class of `T` like this which we can then use for this purpose:
```c++
class RetainableT : public T, public RetainCountable, public Heap::Allocatable<RetainableT>
{
    // May need to explicitly inherit `T` constructors (if we want to use them).
    // for which I guess we can use `using T::T;`
};
```
The `RetainCoutable` adds mechanism to track the ref count, and `Heap::Allocatable` adds `Allocate()` and `Free()` methods. Then we should be able to use `RetainPtr<RetainableT>` .